### PR TITLE
api: Allow MaxSurge to be set on DaemonSets during update

### PR DESCRIPTION
**Pull request by smarterclayton**
MaxSurge allows a daemonset workload to run multiple instances of the same pod on a node during rollout to minimize downtime of the daemonset to other consumers.

This PR adds the feature gate and API for https://github.com/kubernetes/enhancements/issues/1591 and configures round tripping behavior when the gate is enabled or disabled as well as validation rules to prevent inconsistent values. Subsequent PRs will add implementation.

/kind feature
/kind api-change

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/1591-daemonset-surge
```

### CODE REVIEW

File: pkg/apis/apps/types.go [hunk 1]
```diff
@@ -532,19 +532,41 @@ type RollingUpdateDaemonSet struct {
...
+	// the update.
 	// +optional
 	MaxUnavailable intstr.IntOrString
+
+	// The maximum number of nodes with an existing available DaemonSet pod that
+	// can have an updated DaemonSet pod during during an update.
+	// Value can be an absolute number (ex: 5) or a percentage of desired pods (ex: 10%).
+	// This can not be 0 if MaxUnavailable is 0.
+	// Absolute number is calculated from percentage by rounding up to a minimum of 1.
+	// Defaults to 25%.
```
Comment (soltysh) [inline pkg/apis/apps/types.go:554]: You have 25% here, but in `defaults.go` you set 0.

**Review by soltysh — APPROVED** [review]
One nit, the rest lgtm.
/assign @janetkuo @kow3ns @tnozicka

File: pkg/apis/apps/types.go (hunk 1 above)
Comment (smarterclayton) [reply pkg/apis/apps/types.go:554]: 25% definitely would be wierdly inconsistent.  We have to pick a global default anyway so 0.

Comment (smarterclayton) [reply pkg/apis/apps/types.go:554]: Updated to zero - this was copy-paste from the deployment struct.

**Review by tnozicka — COMMENTED** [review]
nit, the rest looks good to me

**Comment by smarterclayton** [comment]
Given timing I'm going to punt this to 1.21 for alpha and first merge.

**Review by janetkuo — APPROVED** [review]

**Comment by smarterclayton** [comment]
Implementation PR is ready for review in https://github.com/kubernetes/kubernetes/pull/96441

**Pull request merged**